eth2/utils/eth2_keystore/src/kdf.rs

```diff
@@ -0,0 +1,203 @@
+    pub fn aes_key(&self) -> &[u8] {
+        &self.0[0..16]
+    }
+}
+
+/// Parameters for `pbkdf2` key derivation.
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Pbkdf2 {
+    pub c: u32,
+    pub dklen: u32,
+    pub prf: Prf,
+    #[serde(serialize_with = "serialize_salt")]
+    #[serde(deserialize_with = "deserialize_salt")]
+    pub salt: Vec<u8>,
+}
+impl Default for Pbkdf2 {
+    fn default() -> Self {
+        let salt = rand::thread_rng().gen::<[u8; SALT_SIZE]>();
+        Pbkdf2 {
+            dklen: DKLEN,
+            c: 262144,
+            prf: Prf::default(),
+            salt: salt.to_vec(),
+        }
+    }
+}
+
+impl Pbkdf2 {
+    /// Derive key from password.
+    pub fn derive_key(&self, password: &str) -> DerivedKey {
+        let mut dk = DerivedKey::zero();
+        let mut mac = self.prf.mac(password.as_bytes());
+        pbkdf2::pbkdf2(&mut mac, &self.salt, self.c, dk.as_mut_bytes());
+        dk
+    }
+}
+
+/// Parameters for `scrypt` key derivation.
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Scrypt {
+    pub dklen: u32,
+    pub n: u32,
+    pub r: u32,
+    pub p: u32,
+    #[serde(serialize_with = "serialize_salt")]
+    #[serde(deserialize_with = "deserialize_salt")]
+    pub salt: Vec<u8>,
+}
+
+/// Compute floor of log2 of a u32.
+fn log2_int(x: u32) -> u32 {
+    if x == 0 {
+        return 0;
+    }
+    31 - x.leading_zeros()
+}
+
+impl Scrypt {
+    pub fn derive_key(&self, password: &str) -> DerivedKey {
+        let mut dk = DerivedKey::zero();
+        // Assert that `n` is power of 2
+        debug_assert_eq!(self.n, 2u32.pow(log2_int(self.n)));
+        let params = scrypt::ScryptParams::new(log2_int(self.n) as u8, self.r, self.p);
+        scrypt::scrypt(password.as_bytes(), &self.salt, &params, &mut dk.0);
+        dk
+    }
+}
+
+impl Default for Scrypt {
+    fn default() -> Self {
+        let salt = rand::thread_rng().gen::<[u8; SALT_SIZE]>();
+        Scrypt {
+            dklen: DKLEN,
+            n: 262144,
+            r: 8,
+            p: 1,
+            salt: salt.to_vec(),
+        }
+    }
+}
+
+/// Serialize `salt` to its hex representation.
+fn serialize_salt<S>(x: &Vec<u8>, s: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    s.serialize_str(&hex::encode(x))
+}
+
+/// Deserialize `salt` from its hex representation to bytes.
+fn deserialize_salt<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+where
+    D: de::Deserializer<'de>,
+{
+    struct StringVisitor;
+    impl<'de> de::Visitor<'de> for StringVisitor {
+        type Value = Vec<u8>;
+        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+            formatter.write_str("String should be hex format")
+        }
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+        where
+            E: de::Error,
+        {
+            hex::decode(v).map_err(E::custom)
```

Comment:
I think we can leave this as-is for now. I'm a little hesitant to add this in, since if the data is malformed I feel like the error will be easier to understand this way as opposed to an invalid password error or something. Maybe I'm wrong here?

Reply:
My thinking was more from the case if someone has exported a keystore from say python which does not guarantee an even number of hex digits as it will trim leading zeros unless explicitly told not to. Then our importation of that keystore would fail.

Reply:
Makes sense, I'll add this :)

Reply:
Just pushed this fix :)